domain/model/state/state.go

```diff
@@ -793,71 +788,178 @@ func (s *State) ListAllModels(ctx context.Context) ([]coremodel.Model, error) {
 
-			model.CloudRegion = cloudRegion.String
-			model.ModelType = coremodel.ModelType(modelType)
-			model.Owner = user.UUID(userUUID)
-			model.Credential = credKey
-			rval = append(rval, model)
+			modelCredential, err := credentialstate.GetCloudCredential(ctx, s, tx, corecredential.ID(stateModel.CredentialID))
+			if err != nil {
+				return fmt.Errorf("cannot get credential %q for model %q: %w", stateModel.CredentialID, stateModel.UUID, err)
+			}
+			modelCredentials[stateModel.UUID] = modelCredential.AsCredential()
 		}
 
-		return rows.Err()
+		return nil
 	})
+	if err != nil {
+		return nil, fmt.Errorf("getting all models: %w", err)
+	}
+
+	hostedModels := make([]coremodel.HostedModel, 0, len(resultsForModel))
+	for _, stateModel := range resultsForModel {
+		coreModel, err := stateModel.toCoreModel()
+		if err != nil {
+			return nil, fmt.Errorf("transforming model %q from db: %w", stateModel.UUID, err)
+		}
+
+		hostedModels = append(hostedModels, coremodel.HostedModel{
+			Model:      coreModel,
+			Cloud:      modelClouds[stateModel.UUID],
+			Credential: modelCredentials[stateModel.UUID],
+		})
+	}
+	return hostedModels, nil
+}
+
+// ModelLastLogins lists all models along with the last login by the specified
+// user. This function will filter on models that have a "life" value in the
+// includeLifes slice. If no matching models exist, an empty slice is returned.
+func (s *State) ModelLastLogins(
+	ctx context.Context,
+	userID user.UUID,
+	includeLifes []corelife.Value,
+) ([]coremodel.ModelWithLogin, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	query := `
+SELECT &dbModel.*, &userModelLastLogin.*
+FROM v_model
+LEFT JOIN model_last_login AS mll
+    ON v_model.uuid = mll.model_uuid
+    AND mll.user_uuid = $M.user_uuid
+WHERE life IN ($lifeList[:])
+`
 
+	inputLifes := lifeList(includeLifes)
+	args := sqlair.M{
+		"user_uuid": userID,
+	}
+
+	stmt, err := sqlair.Prepare(query, dbModel{}, userModelLastLogin{}, inputLifes, args)
+	if err != nil {
+		return nil, fmt.Errorf("preparing select models statement: %w", err)
+	}
+
+	resultsForModel := []dbModel{}
+	lastLogins := []userModelLastLogin{}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, inputLifes, args).GetAll(&resultsForModel, &lastLogins)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return fmt.Errorf("getting models from database: %w", err)
+		}
+		return nil
+	})
 	if err != nil {
 		return nil, fmt.Errorf("getting all models: %w", err)
 	}
 
+	rval := make([]coremodel.ModelWithLogin, 0, len(resultsForModel))
+	for i, stateModel := range resultsForModel {
+		coreModel, err := stateModel.toCoreModel()
+		if err != nil {
+			return nil, fmt.Errorf("transforming model %q from state: %w", stateModel.UUID, err)
+		}
+		rval = append(rval, coremodel.ModelWithLogin{
+			Model:     coreModel,
+			UserID:    user.UUID(lastLogins[i].UserID),
```

Comment:
Maybe use the user UUID from the the argument here (i.e. `userID`) since this lastLogins[i].UserID may be empty if the user never logged in to the model